applications/tari_base_node/src/recovery.rs

```diff
@@ -0,0 +1,166 @@
+        },
+    };
+    let rules = ConsensusManagerBuilder::new(node_config.network.into()).build();
+    let factories = CryptoFactories::default();
+    let validators = Validators::new(
+        FullConsensusValidator::new(rules.clone()),
+        StatelessBlockValidator::new(rules.clone(), factories.clone()),
+    );
+    let db_config = BlockchainDatabaseConfig {
+        orphan_storage_capacity: node_config.orphan_storage_capacity,
+        pruning_horizon: node_config.pruning_horizon,
+        pruning_interval: node_config.pruned_mode_cleanup_interval,
+    };
+    let db = BlockchainDatabase::new(main_db, &rules, validators, db_config).map_err(|e| {
+        error!(target: LOG_TARGET, "Error could not create BlockchainDatabase {}", e);
+        format!("Could not create BlockchainDatabase: {}", e.to_string())
+    })?;
+    do_recovery(db, temp_db).await?;
+
+    info!(
+        target: LOG_TARGET,
+        "Node has completed recovery mode, it will try to cleanup the db"
+    );
+    match &node_config.db_type {
+        DatabaseType::LMDB(p) => {
+            let new_path = Path::new(p).join("temp_recovery");
+            remove_lmdb_database(&new_path).map_err(|e| {
+                error!(target: LOG_TARGET, "Error removing recovery db: {}", e);
+                format!("Could not remove recovery DB: {}", e.to_string())
+            })
+        },
+        _ => {
+            error!(target: LOG_TARGET, "Recovery mode is only available for LMDB");
+            Ok(())
+        },
+    }
+}
+
+// Function to handle the recovery attempt of the db
+async fn do_recovery<D: BlockchainBackend + 'static>(db: BlockchainDatabase<D>, temp_db: D) -> Result<(), String> {
+    // We dont care about the values, here, so we just use mock validators, and a mainnet CM.
+    let rules = ConsensusManagerBuilder::new(NetworkType::LocalNet).build();
+    let validators = Validators::new(MockValidator::new(true), MockValidator::new(true));
+    let temp_db_backend = BlockchainDatabase::new(temp_db, &rules, validators, BlockchainDatabaseConfig::default())
+        .map_err(|e| e.to_string())?;
+    let max_height = temp_db_backend
+        .get_chain_metadata()
+        .map_err(|e| e.to_string())?
+        .height_of_longest_chain
+        .unwrap_or(0);
+    // we start at height 1
+    let mut counter = 1;
+    print!("Starting recovery at height: ");
+    loop {
+        print!("{}", counter);
+        io::stdout().flush().unwrap();
+        trace!(target: LOG_TARGET, "Asking for block with height: {}", counter);
+        let block = temp_db_backend.fetch_block(counter).map_err(|e| e.to_string())?.block;
+        trace!(target: LOG_TARGET, "Adding block: {}", block);
+        db.add_block(Arc::new(block)).map_err(|e| {
+            info!(
+                target: LOG_TARGET,
+                "Stopped recovery at height {}, reason: {}",
+                counter,
+                e.to_string()
+            );
+            e.to_string()
+        })?;
```

Comment:
Use async db so to not block a tokio thread (yes, it's not a big deal in this case but we should get in the habit of doing it)

Use `anyhow!` to provide extra info.

```suggestion
        async_db::add_block(db.clone(), Arc::new(block)).await.map_err(|e| {
            anyhow!(
                "Stopped recovery at height {}, reason: {}",
                counter,
                e
            )
        })?;
```